src/test/compile-fail/issue-23080-2.rs

```diff
@@ -10,6 +10,8 @@
 
 // ignore-tidy-linelength
 
+//~^^^^^^^^^^^^ ERROR
```

Comment:
So the reason our errors get worse in this case is that we end up calling `report_overflow_error` with the dummy spans from the canonicalized query -- except from `traits::project`, [here](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/project.rs#L574), rather than directly from `traits::select`. We enter this codepath via [this line](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/select.rs#L3272) in `select`. The whole `TraitQueryMode` mechanism would've prevented this by propagating the error instead of reporting immediately, but that's only implemented within `select` and not `project`. That's why this situation arises.

Looking at [this comment](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/project.rs#L545-L563), though, it feels like reporting an overflow there is already something of a temporary stopgap. So I don't think it makes sense to invade and refactor all of `project` as well with the `TraitQueryMode` setup just to fix this error. Instead maybe we should just take the hit (I don't think this is anything like a common path anyway). Open to other ideas though.